block/manager.go

```diff
@@ -101,9 +101,28 @@ type Manager struct {
 
 // getInitialState tries to load lastState from Store, and if it's not available it reads GenesisDoc.
 func getInitialState(store store.Store, genesis *cmtypes.GenesisDoc) (types.State, error) {
+	// Load the state from store.
 	s, err := store.GetState()
 	if err != nil {
+		// If the user is starting a fresh chain (or hard-forking), we assume the stored state is empty.
 		s, err = types.NewFromGenesisDoc(genesis)
+		store.SetHeight(s.LastBlockHeight)
+	} else {
+		// Perform a sanity-check to stop the user from
+		// using a higher genesis than the last stored state.
+		// if they meant to hard-fork, they should have cleared the stored State
+		if uint64(genesis.InitialHeight) > s.LastBlockHeight {
+			return types.State{}, fmt.Errorf("genesis.InitialHeight (%d) is greater than last stored state's LastBlockHeight (%d)", genesis.InitialHeight, s.LastBlockHeight)
+		}
+	}
+
+	// The store should not contain a block at genesis.InitialHeight + 1
+	// because InitChain is called at that height, but a block is not built and stored.
+	// this indicates the user didn't properly clear the store when hard-forking.
+	// Perform a sanity-check:
+	_, loadBlockErr := store.GetBlock(s.InitialHeight)
+	if loadBlockErr == nil {
+		return types.State{}, fmt.Errorf("unexpected stored block at genesis.InitialHeight (%d)", s.InitialHeight)
 	}
 	return s, err
```

Comment:
Based on feedback during retro 

the first if condition should check for empty state before loading genesis.
then if there is an error loading genesis we should return
if it was a non empty error, we should just return the error
update the last return to return nil as we should have handled all error cases by then

short hand update example
```go
	// Load the state from store.
	s, err := store.GetState()
	if err != nil && store_is_empty {
		// If the user is starting a fresh chain (or hard-forking), because the stored state is empty, load the genesis doc.
		s, err = types.NewFromGenesisDoc(genesis)
		if err != nil {
		   return err
		 }
		store.SetHeight(s.LastBlockHeight)
	} else if err != nil {
	         return err
	} else {
		// Perform a sanity-check to stop the user from
		// using a higher genesis than the last stored state.
		// if they meant to hard-fork, they should have cleared the stored State
		if uint64(genesis.InitialHeight) > s.LastBlockHeight {
			return types.State{}, fmt.Errorf("genesis.InitialHeight (%d) is greater than last stored state's LastBlockHeight (%d)", genesis.InitialHeight, s.LastBlockHeight)
		}
	}

	// The store should not contain a block at genesis.InitialHeight + 1
	// because InitChain is called at that height, but a block is not built and stored.
	// this indicates the user didn't properly clear the store when hard-forking.
	// Perform a sanity-check:
	_, loadBlockErr := store.GetBlock(s.InitialHeight)
	if loadBlockErr == nil {
		return types.State{}, fmt.Errorf("unexpected stored block at genesis.InitialHeight (%d)", s.InitialHeight)
	}
	return s, nil
```